### PR TITLE
Use URIs in REST endpoints rather the the database id

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

The `VersionId` is supposed to be an implementation detail. Also, it's not possible to read the id for a type when it was already created.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201095311341924/1202642946839413) _(internal)_

## 🔍 What does this change?

- Add a public conversion method to get a version id by URI
- Use URIs in REST GET endpoints
- drive-by: Add error-logging output when REST methods are failing

## ⚠️ Known issues

- Due to the way SQLx implements their `Executor` trait, the functionality for getting a URI by id needs two entrypoints, but can rely on the same internal methods. This will be solved by a follow-up task

## 🛡 What tests cover this?

The test-requests were adjusted to use URL-encoded URIs instead

## ❓ How to test this?

- Install dependencies:

    ```sh
    cargo make yarn
    ```

- Deploy the database and populate it:

    ```sh
    cargo make deployment-up
    cargo make recreate-db
    cargo make migrate-up
    ```

- Run the binary

    ```sh
    cargo make run
    ```

- Run the tests

    ```sh
    cargo make test-rest-api
    ```

## 📹 Demo

image

### CODE REVIEW

File: packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs [hunk 1]
```diff
@@ -642,10 +637,25 @@ impl PostgresDatabase {
             .attach_printable_lazy(|| uri.clone())?
             .get(0))
     }
+
+    #[allow(
+        clippy::same_name_method,
+        reason = "This is required because the way SQLx implements their executors"
+    )]
+    async fn version_id_by_uri(
```
Comment (thehabbos007) [inline packages/graph/hash_graph/lib/graph/src/store/postgres/mod.rs:645]: Sneaky way to get around the pass-by value situation :+1:

File: packages/graph/hash_graph/tests/rest-test.http [hunk 2]
```diff
@@ -19,11 +19,10 @@ Accept: application/json
     client.test("status", function() {
         client.assert(response.status === 200, "Response status is not 200");
     });
-    client.global.set("text_data_type_id", response.body.version_id);
 %}
 
 ### Get Text data type
-GET http://127.0.0.1:3000/data-type/{{text_data_type_id}}
+GET http://127.0.0.1:3000/data-type/https%3A%2F%2Fblockprotocol.org%2F%40blockprotocol%2Ftypes%2Fdata-type%2Ftext%2Fv%2F1
```
Comment (thehabbos007) [inline packages/graph/hash_graph/tests/rest-test.http:25]: May I suggest we do this instead:
```suggestion
GET http://127.0.0.1:3000/data-type/{{encodeURIComponent("https://blockprotocol.org/@blockprotocol/types/data-type/text/v/1")}}
```
could you check if this works with the jetbrains IDE?
It'll make it easier to glance over :)

**Review by thehabbos007 — CHANGES_REQUESTED** [review]
Couple smaller things and something we should try at the end. Thanks!

File: packages/graph/hash_graph/tests/rest-test.http (hunk 2 above)
Comment (TimDiekmann) [reply packages/graph/hash_graph/tests/rest-test.http:25]: Sadly, it does not work. However, it's possible to assign a variable in the statement above:
```
> {%
    client.test("status", function() {
        client.assert(response.status === 200, "Response status is not 200");
    });
    client.global.set("text_data_type_uri", encodeURIComponent(response.body.inner["$id"]));
%}

### Get Text data type
GET http://127.0.0.1:3000/data-type/{{text_data_type_uri}}

> {%
    client.test("status", function() {
        client.assert(response.status === 200, "Response status is not 200");
    });
%}
```

Comment (thehabbos007) [reply packages/graph/hash_graph/tests/rest-test.http:25]: Okay, I think we should do that for the test that previously depended on the globals to be set

Comment (TimDiekmann) [reply packages/graph/hash_graph/tests/rest-test.http]: Applied in https://github.com/hashintel/hash/pull/838/commits/8fcb33ee8372ec111a8fb05ff99cec334672c2da

**Review by thehabbos007 — APPROVED** [review]
Beautiful!

**Pull request merged**